source/mainnet/smart-contracts/guides/unit-test-contract.rst

```diff
@@ -339,18 +339,144 @@ state and balance fields:
+    ) -> Result<(), Error> {
+        let sender = ctx.sender();
+
+        // Get balance for the sender, or reject if the sender is not found or the
+        // balance is zero.
+        let sender_balance = match host.state().balances.get(&sender) {
+            Some(bal) if *bal > Amount::zero() => *bal,
+            _ => return Err(Error::WithdrawWithoutFunds),
+        };
+
+        match sender {
+            Address::Account(acc) => host.invoke_transfer(&acc, sender_balance)?,
+            Address::Contract(addr) => {
+                let entrypoint: OwnedEntrypointName = ctx.parameter_cursor().get()?;
+                // At this point we are handing out the control out to an unknown
+                // smart contract. This contract can call this entry point
+                // again multiple times before the rest of the code is reached.
+                host.invoke_contract(
+                    &addr,
+                    &Parameter(&[]),
+                    entrypoint.as_entrypoint_name(),
+                    sender_balance,
+                )?;
+            }
+        };
+
+        // Reset the sender's balance to zero.
+        // This code is reached only after transfering CCD back/calling an
+        // external contract.
+        if let Some(mut v) = host.state().balances.get_mut(&sender) {
+            *v = Amount::zero();
+        }
+
+        Ok(())
+    }
+
+The problem in the code above is that resetting the sender's balance to zero happens *after* the call to an external contract is completed.
+The sender's balance in the *contract state* is used to determine how much funds should be transferred to the sender.
+Since it is not updated, the external contract can make a call back to ``reentrant_withdraw`` and pass the balance validation.
+Testing this behavior with mocks require some insights.
+In particular, we are going to mimic the original ``reentrant_withdraw`` code in the mock entrypoint.
+
+.. code-block:: rust
+
+    #[concordium_test]
+    fn test_reentrant_withdraw() {
+        ...
+
+        // Assume that `CONTRACT_ADDRESS` has 1 micro CCD
+        // Set the contract balance to 2 micro CCD
+        host.set_self_balance(Amount::from_micro_ccd(2));
+
+        // Set up a mock entrypoint that calls back to our contract.
+        // We emulate the `reentrant_withdraw` logic here.
+        host.setup_mock_entrypoint(
+            CONTRACT_ADDRESS,
+            OwnedEntrypointName::new_unchecked("reentrant_withdraw".to_string()),
+            MockFn::new_v1(|_parameter, _amount, balance, state: &mut State<_>| {
+                // We cannot call `invoke_contract` inside this mock, but we have
+                // access to the `balance`, which is the balance of the contract
+                // making this invocation. So we can simulate invoking withdraw by
+                // subtracting the sender's amount stored in the contract state
+                // from the balance.
```

Comment:
Perhaps we could mention that we are simulation a reentrancy attack, where we will withdraw the balance twice.

Reply:
I added a line about this before the mock entrypoint definition.